### PR TITLE
Fix appveyor exit code handling

### LINKED ISSUE #1940: Move tests
After the big split, we can move the tests to the individual `packages`.   That way each test can be run from that package, or from the top level using `lerna run test`.   We are hoisting dev dependencies, meaning that karma and mocha would only be installed at the root of the repo.

### LINKED ISSUE #3477: jupyterlab-0.31.0rc1 starts with a blank page
when I launch jupyterlab-0.31.0rc1, it opens a blank page:
jupyterlabindex2

with  jupyterlab-0.30.6 , I had this error
jupyterlabindex

What may I have done wrong ? (window-10, Python-3.6.4, not anaconda)

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #3477.   Fixes #1940. Refactors the tests to run independently by package and fixes race conditions that were causing intermittent test failures.

- Cleans up terminal reconnect behavior
- Cleans up kernel restart behavior

The original issue with appveyor is that we were not handling the exit codes of each of the entries in `scripts/appveyor.cmd`, nor the final exit code.  We don't have an option of using something like `set -e` from bash, so we have to explicitly test the exit status each time.

### CODE REVIEW

**Comment by jasongrout** [comment]
When convenient, could you give a short explanation of what the problem in #3477 was?

**Comment by blink1073** [comment]
Updated.  This refactor has also been something I've been meaning to do since the lerna switch.

**Comment by blink1073** [comment]
I was finally able to reproduce the failing test locally by using a python 2.7 environment, much easier than poking at appveyor over and over...

**Comment by blink1073** [comment]
Finally done!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -124,6 +124,39 @@ jlpm run build:test
 jlpm test
 ```
 
+You can run tests for an individual package by changing to the appropriate
```
Comment (ellisonbg) [inline CONTRIBUTING.md]: Thanks this is very helpful!

File: buildutils/src/create-test-package.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      process.stderr.write(msg);
+      process.exit(1);
+  }
+  let name = process.argv[2];
+  let pkgPath = path.resolve(path.join('.', 'packages', name));
+  if (!fs.existsSync(pkgPath)) {
+    console.error('Package does not exist: ', name);
+    process.exit(1);
+  }
+  let dest = path.resolve(`./tests/test-${name}`);
```
Comment (ellisonbg) [inline buildutils/src/create-test-package.ts]: should `path.join` be used here?

File: jupyterlab/tests/test_app.py [hunk 3]
```diff
@@ -122,5 +129,80 @@ def _process_finished(self, future):
...
 
+class KarmaTestApp(TestApp):
+    """A notebook app that runs the jupyterlab karma tests.
+    """
+    karma_pattern = Unicode('src/*.spec.ts')
+    karma_base_dir = Unicode('')
+
+    def get_command(self):
+        """Get the command to run."""
+        terminalsAvailable = self.web_app.settings['terminals_available']
```
Comment (ellisonbg) [inline jupyterlab/tests/test_app.py]: does this setting still make sense now that terminals work on Windows?

File: buildutils/src/create-test-package.ts (hunk 2 above)
Comment (blink1073) [reply buildutils/src/create-test-package.ts]: Backslash paths are still valid in Windows.

File: jupyterlab/tests/test_app.py (hunk 3 above)
Comment (blink1073) [reply jupyterlab/tests/test_app.py]: Support for terminals depends on the not-yet-released Notebook 5.3

File: buildutils/src/create-test-package.ts (hunk 2 above)
Comment (ellisonbg) [reply buildutils/src/create-test-package.ts]: +1

File: jupyterlab/tests/test_app.py (hunk 3 above)
Comment (ellisonbg) [reply jupyterlab/tests/test_app.py]: Ahh, got it!

**Comment by ellisonbg** [comment]
Needs rebase, otherwise looks good!

**Pull request merged**